src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +28,132 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn wake_after(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+type AwaitingCallerId = usize;
+type WokeUpCallersIds = Vec<AwaitingCallerId>;
+
+#[derive(Debug)]
+struct AwaitingCaller {
+    id: AwaitingCallerId,
+    time_to_wake_up: Duration,
+    waker_tx: Option<oneshot::Sender<()>>,
+}
+
+#[derive(Debug, Default)]
+struct SystemTimeSourceStubState {
+    t: Option<DateTime<Utc>>,
+    next_caller_id: AwaitingCallerId,
+    awaiting_callers: Vec<AwaitingCaller>,
+}
+
 #[derive(Debug, Clone)]
 pub struct SystemTimeSourceStub {
-    t: Arc<Mutex<Option<DateTime<Utc>>>>,
+    state: Arc<Mutex<SystemTimeSourceStubState>>,
 }
 
 impl SystemTimeSourceStub {
     pub fn new() -> Self {
         Self {
-            t: Arc::new(Mutex::new(None)),
+            state: Arc::new(Mutex::new(SystemTimeSourceStubState::default())),
         }
     }
 
     pub fn new_set(t: DateTime<Utc>) -> Self {
+        let state = SystemTimeSourceStubState {
+            t: Some(t),
+            next_caller_id: 1,
+            awaiting_callers: vec![],
+        };
+
         Self {
-            t: Arc::new(Mutex::new(Some(t))),
+            state: Arc::new(Mutex::new(state)),
         }
     }
 
     pub fn set(&self, t: DateTime<Utc>) {
-        *self.t.lock().unwrap() = Some(t);
+        let mut state = self.state.lock().unwrap();
+
+        state.t = Some(t);
     }
 
     pub fn unset(&self) {
-        *self.t.lock().unwrap() = None;
+        let mut state = self.state.lock().unwrap();
+
+        state.t = None;
+    }
+
+    pub fn advance(&self, time_quantum: Duration) -> WokeUpCallersIds {
+        let mut state = self.state.lock().unwrap();
+        let mut woke_up_callers_ids = vec![];
+
+        state.awaiting_callers.retain_mut(|awaiting_caller| {
```

Comment:
If I understood correctly, this is scanning the entire list of awakenings to apply the predicate + it touches the element. What if we consider implementing it in the form of a prioririty queue based on binary heap? We would get O(1) sorted access.

Reply:
fixed, thanks for such good comment